### PR TITLE
[CLEANUP] Removed duplicate form validation on signup

### LINKED ISSUE #1057: Unify form validation between client and server
(This is a refactoring/cleanup task.)

---

Right now, form validation code is duplicated between client and server.

For example, user signup is validated here on the server:

https://github.com/Felienne/hedy/blob/c778f7c221e7721e725460db3b49a4df006b72bc/website/auth.py#L124-L133

And the same validations are performed here on the client:

https://github.com/Felienne/hedy/blob/c778f7c221e7721e725460db3b49a4df006b72bc/static/js/auth.js#L46-L52

This duplication leads to issues (for example #1026) where the logic between the two doesn't agree; also the maintenance cost of these two sets of rules is increased. 

There is one difference, and that is that the client has information about WHERE to display the error message, and the client localizes the error message.

It would be better if the server was the only component validating, and it replied with enough information that the client could parse the response and show the same, localized, appropriate error messages highlighting the appropriate fields.

### LINKED ISSUE #1555: [BUG] Users are unable to empty optional profile fields after setting a value
**Describe the bug**
Currently when a users updates their profile with an optional field everything works fine. However, when trying to remove this set field (by leaving it empty in the update form) this change won't be updated in the database as well. This is due to the structure of our back-end and we should write an additional function to verify if currently found fields in the database are still in the updated inputs, if not: delete them from the database.

**Expected behavior**
When users leave an earlier saved optional field empty the old value should be correctly deleted from the database.

### PR DESCRIPTION
**Pull request by TiBiBa**
**Description**
As pointed out by @rix0rrr in #1057 we currently perform some duplicate form validations on signup. In this PR we assign all validations to the back-end, returning specific errors when a given value is incorrect. On the front-end we get the corresponding message to the error in the requested language. When, for some reason, the error is unknown on the front-end we return a general error message. We also improve the form handling with empty values, enabling users to empty previously set optional field such as birth year, country and programming experience.

**Fix for**
This PR fixed #1057 and fixed #1555.

**How to test**
As this is a cleanup task no noticeable changes are made for the user.

**Checklist**

If you're unsure about any of these, don't hesitate to ask. We're here to help!
- [ ] Links to an existing issue or discussion (if not, create an issue first)
- [ ] Describes changes clear in the format above (present tense, no subject)
- [ ] Has a "how to test" section
- [ ] Only one thing is done in this pull request (specifically please try to refrain from mixing textual changes to the yamls from code changes)

### CODE REVIEW

File: static/js/auth.ts
Comment (rix0rrr) [inline static/js/auth.ts:190]: What if this returns something that's not in the dictionary though?

File: website/auth.py
Comment (rix0rrr) [reply website/auth.py:154]: Also, I would probably do some structured data like `{ code: 'not_object', message: 'body must be an object' }`.

**Review by rix0rrr — COMMENTED** [review]
Thanks for picking this up! Makes me feel a lot better. I have some questions though.

File: website/auth.py
Comment (TiBiBa) [reply website/auth.py:154]: Didn't think about it that way! As the translations are already available to the template at the front-end I thought this would be the most straight-forward implementation. I agree that it would be better to return the translated error message from the back-end, but we would have to include the translated files to `auth.py`

File: templates/signup.html
Comment (TiBiBa) [reply templates/signup.html:32]: Indeed! Will fix right away

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts:130]: My reasoning was as follows: If we return a 500 it doesn't make any sense to give the user the option to create an account as we have some server issues. So creating an account probably won't working any way. We return a 400 if either the username or password is an invalid type, in any other case we return a 403. In practice users will never see a 400 as the inputs are always parsed as a string and the object is always a dict.

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts:165]: I thought that if the timeout was a bit longer users would be able to read the "success" message before the screens reloads. The reason we reload is because if a users changes their preferred language we want to update the complete UI, including the menubar etc. As the language is stored in the session the easiest way to do this is through a reload. 

It might be best to improve this implementation to only reload if the preferred language is changed. We can simply add this as a boolean to the result of the back-end. Will take a look!

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts:190]: This is a mistake! I had in mind to verify if the value exists, but completely forgot to implement this. We have to add an additional check if the value exists in the dict. The other solution might be as you suggested in the first comment to sent the already translated error message from the back-end.

File: website/auth.py
Comment (rix0rrr) [reply website/auth.py:154]: We should be sticking the translations on the global `g` object, so they're straight up available in the language we need, like this:

https://github.com/Felienne/hedy/blob/34aec748748cb1c30be932b04a26f415f00656f1/app.py#L280

It's a bit unfortunate the "auth" messages have their own section, so now we need to do hulahoops to get the information in the right place, but so be it...

I can see how you might just want to merge what you have today though, so you can also disregard what I'm saying and merge what you have. It's already better than what it used to be.

File: static/js/auth.ts
Comment (rix0rrr) [reply static/js/auth.ts:165]: A second is enough if you know where to look, but not all users might know that.

Another solution is to persist the success message in a place where it would survive a reload.

**Comment by rix0rrr** [comment]
As stated above -- you're improving the state of the world with this PR, so I don't want to be the one blocking it. So if you feel like merging, please go ahead. Nothing in here that would break anything, as far as I can see.

On the other hand, I do have some... *cough* suggestions *cough* around user experience that I made above you might consider 😜

File: website/auth.py
Comment (TiBiBa) [reply website/auth.py]: Just re-wrote all translations to be created on the back-end. In the case of no response text given we still get the `ajax_error` in the front-end as fallback.

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts]: Just changed it to `1500`. I agree that this is not an ideal approach, but it should be suitable for now. I will open an issue to think about a more elegant way to implement the feedback and reload the page.

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts]: Fixed by no longer be dependent on the error-code. We simply use `auth.error` to show the message from the back-end. We have an `else` to catch any unexpected behavior and then show the `ajax_error`.

**Comment by TiBiBa** [comment]
> As stated above -- you're improving the state of the world with this PR, so I don't want to be the one blocking it. So if you feel like merging, please go ahead. Nothing in here that would break anything, as far as I can see.
> 
> On the other hand, I do have some... _cough_ suggestions _cough_ around further improvements you could make that you might consider 😜

I agree! Just finished shifting the translations from the front-end to the back-end, will do some testing know but it would be great if you could take a look. Will change the PR to open if I'm ready!

File: static/js/auth.ts
Comment (TiBiBa) [reply static/js/auth.ts]: Fixed this by always showing the button as in the original implementation.

**Comment by TiBiBa** [comment]
Almost forgot: Because we switch all validations to the back-end we have to re-write (a lot!) from the e2e tests. As soon as the tests pass this is ready to merge!

**Comment by TiBiBa** [comment]
There is still one e2e test failing, but I have no idea why. For the `test_invalid_change_password` function it works fine when giving different invalid bodies. But when giving a correct body but an incorrect current password it results in a 500 error code. Any prints within the function don't work.

File: website/auth.py
Comment (rix0rrr) [inline website/auth.py:359]: @TiBiBa I found it.

```suggestion
            return g.auth_texts.get('password_invalid'), 403
```

Here's what I did to find this issue:

## Run a regular Hedy app in one terminal tab

```
$ python app.py
```

Note that it runs on port 8080.

## Run the failing test by itself in another terminal tab

```
$ env ENDPOINT=http://localhost:8080 pytest -q tests_e2e.py -k test_invalid_change_password
```

I mostly copy/pasted this line from `build-tools/github/validate-e2e`, filled in the `$PORT` with `8080`, and added `-k test_invalid_change_password` to *only* run the specific test that was failing.

## Notice the error in the app error output

Because this makes sure that the app logging output is only relevant to the test at hand (and not all other 30+ successful tests), it's now much easier to see the following error in the scrollback:

```
INFO:werkzeug:127.0.0.1 - - [16/Jan/2022 11:19:39] "POST /auth/change_password HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/rix0rrr/Dev/hedy/.env/lib/python3.8/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)

  (...bunch of irrelevant Flask innards...)

  File "/Users/rix0rrr/Dev/hedy/website/auth.py", line 125, in inner
    return f(current_user(), *args, **kws)
  File "/Users/rix0rrr/Dev/hedy/website/auth.py", line 360, in change_password
    return g.auth.texts.get('password_invalid'), 403
  File "/Users/rix0rrr/Dev/hedy/.env/lib/python3.8/site-packages/werkzeug/local.py", line 347, in __getattr__
    return getattr(self._get_current_object(), name)
AttributeError: '_AppCtxGlobals' object has no attribute 'auth'
```

Showing me that the error happens on:

```py
    return g.auth.texts.get('password_invalid'), 403   # AttributeError: '_AppCtxGlobals' object has no attribute 'auth'
```

And then looking at the corresponding code I can see it's a typo that needs to be `g.auth_texts`.

**Review by rix0rrr — APPROVED** [review]
Good to go once you get rid of all the debugging `print` statements 😋

File: website/auth.py
Comment (TiBiBa) [reply website/auth.py:359]: Wow good catch! Thanks!

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**
